torch/nn/parallel/_functions.py

```diff
@@ -1,11 +1,22 @@
 import warnings
 
 import torch
+import types
+import importlib
 from . import comm
 from torch.autograd import Function
-from torch._utils import _get_device_index
-from typing import List, Optional
-
+from torch._utils import _get_device_index, _get_available_device_type
+from typing import List, Optional, Tuple
+from functools import lru_cache
+
+@lru_cache(maxsize=None)
+def _get_device_impl() -> Tuple[str, types.ModuleType]:
+    ddp_device_type = _get_available_device_type()
```

Comment:
IIUC, in `_get_available_device_type`, CUDA will overshadow XPU if both are available. Does this mean, we wanna use XPU, we will need to 1) use PyTorch with CPU-only build 2) load XPU lib dynamically?

Reply:
For 1), the CPU-only build is not a MUST option. As long as pytorch is installed, and CUDA is not available, for example, no CUDA runtime or device, XPU can be loaded.
For 2), Yes. XPU will be import by user model script, and register to PyTorch.